### PR TITLE
 Add utility functions to detect and select local lambda functions.

### PR DESCRIPTION
**Pull request by mpiroc**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [ ] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

Detect lambda functions from local `template.y[a]ml` files, and provide the ability to select from detected functions. Note that detection directly from source code is not part of this changeset--@awschristou is investigating the best way to do this.

## Motivation and Context

Several tasks this sprint (related to local debugging of node lambdas) depend on the ability to detect lambda functions in the current workspace, and select from detected lambda functions. In order to unblock parallel implementation of these tasks, I'm making this a separate PR.

## Related Issue(s)

Primary: #130

Also related: #131 and #134

## Testing

The new functionality is not yet exposed to the user, so it's currently tested by unit tests only. CodeCoverage [reports](https://codecov.io/gh/aws/aws-toolkit-vscode/compare/develop...pirocchi%2Fselect-template/diff) 97.5% coverage of new/changed code.

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/lambda/local/detectLocalLambdas.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    const resources = template.Resources
+    if (!resources) {
+        return []
+    }
+
+    return Object.keys(resources)
+        .filter(key => resources[key].Type === 'AWS::Serverless::Function')
+        .map(key => ({
+            label: key,
+            description: templatePath,
```
Comment (awschristou) [inline src/lambda/local/detectLocalLambdas.ts]: Why did you add `label` and `description` ? They aren't in `LocalLambda` , if callers are expecting those fields, they should be in the interface.

File: src/test/lambda/local/detectLocalLambdas.test.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    test('detects no lambdas when workspaceFolders is empty', async () => {
+        const actual = await detectLocalLambdas([])
+
+        assert.ok(actual)
+        assert.equal(actual.length, 0)
+    })
+
+    test('detects no lambdas when template.y[a]ml does not exist', async () => {
+        const workspacePath = await createTemporaryDirectory('vsctk1')
```
Comment (awschristou) [inline src/test/lambda/local/detectLocalLambdas.test.ts]: Since you have a reasonable amount of tests making then tearing down temp folders, consider moving that to a "before/after test" declaration. It keeps the test focused on the actual test, which is easier to read, and you don't have to make sure you properly set up/tear down within each test. You can see an example at the suite level [here](https://github.com/aws/aws-toolkit-vscode/blob/1320413f73096ba705fbf951848ece730e5894da/src/test/filesystem.test.ts#L18)

File: src/lambda/local/detectLocalLambdas.ts (hunk 1 above)
Comment (mpiroc) [reply src/lambda/local/detectLocalLambdas.ts]: Good catch, this is a carryover from the previous iteration. Will move the introduction of these properties to `selectLocalLambda.ts`.

File: src/test/lambda/local/detectLocalLambdas.test.ts (hunk 2 above)
Comment (mpiroc) [reply src/test/lambda/local/detectLocalLambdas.test.ts]: `beforeEach`/`afterEach` are only available in BDD-style tests. I think that switching to BDD has many advantages, but it should be a separate PR.

Comment (mpiroc) [reply src/test/lambda/local/detectLocalLambdas.test.ts]: Update: Chris pointed out that they are available as `setup` and `teardown`. Updating.

File: src/test/lambda/local/util.ts [hunk 3]
```diff
@@ -7,8 +7,25 @@
 
 import * as os from 'os'
 import * as path from 'path'
+import { Uri, WorkspaceFolder } from 'vscode'
 import * as filesystem from '../../../shared/filesystem'
 
+export async function createWorkspaceFolder(prefix: string): Promise<{
```
Comment (awschristou) [inline src/test/lambda/local/util.ts]: `prefix` isn't used in the method

Comment (awschristou) [reply src/test/lambda/local/util.ts]: Did you intend to pass it to `createTemporaryDirectory`?

Comment (mpiroc) [reply src/test/lambda/local/util.ts]: Yep, good catch.
